src/core/lib/security/context/connection_auth_context.h

```diff
@@ -0,0 +1,134 @@
+//
+//
+// Copyright 2024 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#ifndef GRPC_SRC_CORE_LIB_SECURITY_CONTEXT_CONNECTION_AUTH_CONTEXT_H_
+#define GRPC_SRC_CORE_LIB_SECURITY_CONTEXT_CONNECTION_AUTH_CONTEXT_H_
+
+#include <grpc/support/port_platform.h>
+#include <grpc/grpc_security.h>
+
+#include <stddef.h>
+#include <cstdint>
+#include <vector>
+
+#include "src/core/util/orphanable.h"
+
+namespace grpc_core {
+
+class ConnectionAuthContext;
+
+namespace auth_context_detail {
+
+template <typename T>
+GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION void DestroyAuthProperty(void* p) {
+  T::Destroy(static_cast<typename T::ValueType*>(p));
+}
+
+// Tracks all registered/known auth property types (these should only be
+// registered via AuthPropertyTraits at static initialization time).
+class BaseAuthPropertiesTraits {
+ public:
+  // Count of number of registered auth context properties.
+  static uint16_t NumAuthProperties() {
+    return static_cast<uint16_t>(RegisteredTraits().size());
+  }
+
+  // Number of bytes required to store pointers of all registered auth context
+  // properties.
+  static size_t Size() { return NumAuthProperties() * sizeof(void*); }
+
+  // Call the registered destruction function for a context.
+  static void Destroy(uint16_t id, void* ptr) {
+    if (ptr == nullptr) return;
+    RegisteredTraits()[id](ptr);
+  }
+
+  // Allocate a new context id and register the destruction function.
+  template <typename T>
+  static uint16_t AllocateId() {
+    auto& traits = RegisteredTraits();
+    const uint16_t id = static_cast<uint16_t>(traits.size());
+    traits.push_back(DestroyAuthProperty<T>);
+    return id;
+  }
+
+ private:
+  // Allocate a new context id and register the destruction function.
+
+  static std::vector<void (*)(void*)>& RegisteredTraits() {
+    static std::vector<void (*)(void*)> registered_traits;
+    return registered_traits;
+  }
+};
+
+}  // namespace auth_context_detail
+
+class ConnectionAuthContext final : public Orphanable {
```

Comment:
I don't think this needs to be scoped to just AuthContext - I think it'd be perfectly fine here to call this a ConnectionContext and allow usage for non-auth-related things too (and I'd prefer that so that we don't get tempted to invent new mechanisms for such data)

Reply:
Renamed it to ConnectionContext and moved the file to src/core/lib/surface